main.cpp

```diff
@@ -1 +1,411 @@
+        if (((GAME_ARR[0][2] == userToCheck) && (GAME_ARR[1][2] == userToCheck)) ||
+            ((GAME_ARR[2][0] == userToCheck) && (GAME_ARR[2][1] == userToCheck)))
+            return true;
+        else return false;
+    } else return false;
+}
+
+/**
+ * Check if the user given had win
+ * :this method can navigate every size of array (searching for 3 case checked)
+ * @param userToCheck the user value to check
+ * @return true if he win else false
+ */
+bool checkWinR(int userToCheck) {
+    //iterate on each case of the array
+    for (int x = 0; x < GAME_ARR.size(); x++) {
+        for (int y = 0; y < GAME_ARR[0].size(); y++) {
+            //if the case is checked by the user we want to win
+            if (GAME_ARR[x][y] == userToCheck) {
+                //check bounderies for have no error
+                bool allowXplus1 = x + 1 - (int) GAME_ARR.size() < 0;
+                bool allowXplus2 = x + 2 - (int) GAME_ARR.size() < 0;
+                bool allowYplus1 = y + 1 - (int) GAME_ARR[x].size() < 0;
+                bool allowYplus2 = y + 2 - (int) GAME_ARR[x].size() < 0;
+                bool allowYsub2 = y - 2 >= 0;
+                bool allowYsub1 = y - 1 >= 0;
+
+                bool finalBool = false;
+                if (allowXplus1 && allowXplus2) {
+                    //check if a column of 3 case is checked
+                    finalBool = (finalBool ||
+                                 (GAME_ARR[x + 1][y] == userToCheck && GAME_ARR[x + 2][y] == userToCheck));
+                    if (allowYplus1 && allowYplus2) {
+                        //check if a diagonal of 3 case is checked
+                        finalBool = (finalBool ||
+                                     (GAME_ARR[x + 1][y + 1] == userToCheck && GAME_ARR[x + 2][y + 2] == userToCheck));
+                    }
+                    if (allowYsub1 && allowYsub2) {
+                        //check if a diagonal of 3 case is checked
+                        finalBool = (finalBool ||
+                                     (GAME_ARR[x + 1][y - 1] == userToCheck && GAME_ARR[x + 2][y - 2] == userToCheck));
+                    }
+                }
+                if (allowYplus1 && allowYplus2) {
+                    //check if a line of 3 case is checked
+                    finalBool = (finalBool ||
+                                 (GAME_ARR[x][y + 1] == userToCheck && GAME_ARR[x][y + 2] == userToCheck));
+                }
+                if (finalBool) {
+                    //if 3 case checked return true for win
+                    return true;
+                }
+            }
+        }
+    }
+
+    //if no 3 case is found not win so return false
+    return false;
+}
+
+/**
+ * Find the two mid case nearest of the given corner
+ *
+ * @param corner the corner to get mid 1 and mid 2
+ * @return a array containing mid 1 and mid 2 coordinates ([x1, y1, x2, y2])
+ */
+std::array<int, 4> optimalAlgo_findMid1Mid2(std::array<int, 2> corner) {
```

Comment:
Sorry, didn't have the courage to go through your optimal algorithm here... But it seems to go into an endless loop under some circumstances.